modules/partners/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+}
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_proxy::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Currency type for transfer of currencies
+		type Currencies: MultiCurrency<Self::AccountId, CurrencyId = CurrencyId, Balance = Balance>;
+
+		/// Permission to do admin calls in pallet
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Native CurrencyId
+		#[pallet::constant]
+		type GetNativeCurrencyId: Get<CurrencyId>;
+
+		/// Fee for registering a partner_id to account
+		#[pallet::constant]
+		type RegisterFee: Get<Balance>;
+
+		/// The partner's module id
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Number of blocks a referral lasts
+		#[pallet::constant]
+		type ReferralExpire: Get<Self::BlockNumber>;
+
+		/// Treasury's AccountId
+		#[pallet::constant]
+		type TreasuryAccount: Get<Self::AccountId>;
+
+		/// Max size of metadata string
+		#[pallet::constant]
+		type MaxMetadataLength: Get<u32>;
+
+		/// Provides current blocknumber
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = Self::BlockNumber>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Sub Account cant be generated, should not happen
+		SubAccountGenerationFailed,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Referral succesfully set to PartnerId
+		ReferralSet { who: T::AccountId, partner_id: PartnerId },
+		/// New Partner successfully registered
+		PartnerRegistered {
+			owner: T::AccountId,
+			partner_id: PartnerId,
+			partner_account: T::AccountId,
+		},
+		/// Partner's metadata is updated
+		PartnerMetadataUpdated { partner_id: PartnerId },
+	}
+
+	#[pallet::storage]
+	pub type NextId<T: Config> = StorageValue<_, PartnerId, ValueQuery>;
+
+	#[pallet::storage]
+	pub type PartnerMetadata<T: Config> =
+		StorageMap<_, Identity, PartnerId, BoundedVec<u8, T::MaxMetadataLength>, OptionQuery>;
+
+	#[pallet::storage]
+	pub type Referral<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, ReferralInfo<T::BlockNumber>, OptionQuery>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn register_partner(
+			origin: OriginFor<T>,
+			metadata: BoundedVec<u8, T::MaxMetadataLength>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// transfer registration fee to treasury
+			let treasury_account = T::TreasuryAccount::get();
+			<T as module::Config>::Currencies::transfer(
+				T::GetNativeCurrencyId::get(),
+				&who,
+				&treasury_account,
+				T::RegisterFee::get(),
+			)?;
+
+			Self::feeless_register_partner(who, metadata)
+		}
+
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn set_referral(origin: OriginFor<T>, partner: PartnerId) -> DispatchResult {
```

Comment:
Could find a way to make this feeless? Users would be more likely to sign transaction if it were free